### DISCUSSION TITLE
Allow dlopen/dlsym to fail without aborting

### DISCUSSION THREAD

**Discussion by disarray2077** [emscripten-core/emscripten#16254] [Ideas]
Hello! This is my first time interacting with emscripten so sorry If this idea is unreasonable, but with the knowledge I currently have I think it could be a good idea to allow `dlopen`/`dlsym` to fail without aborting when linking support isn't necessary, maybe through some flag like `-s USE_MODULES=0` or something more specific like `-s NO_MODULE="['module.so']"
Currently, I can edit the JS to achieve the same result but doing this is less than ideal.

**Comment by sbc100** [comment]
I think your request is reasonable.  We should probably just return an error code as you say.  If you would like to send a patch that seems reasonable.

I think we may want to continue to log a useful error message at least in debug builds, since calling dlopen or dlsym in builds that don't support it doesn't make much sense and is probably something that the developer would want to fix.  Out of interest, is this not something you would not want to fix your application?   Do you have some reason to want to call those function even though you know they will fail?

**Reply by disarray2077** [reply]
>I think your request is reasonable. We should probably just return an error code as you say. If you would like to send a patch that seems reasonable.

Do you mean a pull request or literally sending a git patch here? I could try.

>Out of interest, is this not something you would not want to fix your application? Do you have some reason to want to call those function even though you know they will fail?

My project uses a closed-source shared library which does not currently support wasm ([Bass](https://www.un4seen.com/)), but the programming language I'm using ([BeefLang](https://www.beeflang.org/)) always tries to load the library with dlopen/dlsym before my code runs, the only thing I can do is handle the failure, but I don't have a chance to handle the failure if the execution is aborted.
Well, I could theoretically rewrite the library bindings to not rely on the language to load everything, but that would be quite troublesome and would create a lot of code useful just for this specific case.

**Reply by sbc100** [reply]
Sorry, I meant open a PR.
